### PR TITLE
Branks42 Engagement Detail Changes

### PR DESCRIPTION
**Pull request by branks42**
***Work Completed***
- `user_response` on EngagementDetail model is now `user_responses`
- New method of validating responses
- Update serializers and views
- New key in payload, `proceed` tells the UI whether to go to the next step or not
- New shape of `user_responses`:
    ```python
        "user_responses": [
            {
                "inputs": [
                    {
                        "userInput": "1989-01-01",
                        "stepInputID": "b0545fac-c400-4bca-a204-84bef0baeeda",
                        "stepInputUIIdentifier": "brake_fast"
                     },
                     {
                         "userInput": "Orange",
                         "stepInputID": "997da4ce-2073-469c-a992-47d9ce19ac4b",
                         "stepInputUIIdentifier": "eggs_and_jelly"
                      }
                  ]
              }, ...]
    ```
- Update tests

***Steps Required***
- New migration

### CODE REVIEW

File: django_workflow_system/api/serializers/user/workflows/engagement_detail.py [hunk 1]
```diff
@@ -140,29 +140,29 @@ def getattr_patched(attr_name):
             return data
 
         if has_required_inputs:  # then having no answers is bad
```
Comment (eikonomega) [inline django_workflow_system/api/serializers/user/workflows/engagement_detail.py:142]: I think we currently have two potential ways that an input can be marked as `required`. As an attribute on the model, like this code is referencing, or as part of the JSON schema for the input itself.

Something to discuss, but may not have time to change.

File: django_workflow_system/api/serializers/user/workflows/engagement_detail.py [hunk 2]
```diff
@@ -140,29 +140,29 @@ def getattr_patched(attr_name):
...
+            if not user_responses:
                 return data
-            if "questions" not in user_response:
+            if "inputs" not in user_responses[-1]:
                 return data
 
         # Ensure all required attributes are present for each question in the payload.
         answer_dict = {}
-        for answer in user_response["questions"]:
+        for answer in user_responses[-1]["inputs"]:
```
Comment (eikonomega) [inline django_workflow_system/api/serializers/user/workflows/engagement_detail.py]: For the sake of completeness, this should probably check not just the last set of provided answers, but all sets of provided answers.

File: django_workflow_system/api/views/user/workflows/engagement_detail.py [hunk 3]
```diff
@@ -167,6 +171,13 @@ def post(self, request, id):
             )
             data = serializer.data
             data["state"] = engagement_serializer.data["state"]
+            # Check if we are able to proceed to the next step
```
Comment (eikonomega) [inline django_workflow_system/api/views/user/workflows/engagement_detail.py:174]: Looks like the location of `proceed` is not documented in the docstring?

**Review by eikonomega — CHANGES_REQUESTED** [review]

File: django_workflow_system/api/views/user/workflows/engagement_detail.py (hunk 3 above)
Comment (branks42) [reply django_workflow_system/api/views/user/workflows/engagement_detail.py:174]: Good catch!

File: django_workflow_system/api/serializers/user/workflows/engagement_detail.py (hunk 2 above)
Comment (branks42) [reply django_workflow_system/api/serializers/user/workflows/engagement_detail.py]: Good call!

**Pull request merged**